vllm/compilation/sequence_parallelism.py

```diff
@@ -0,0 +1,289 @@
+# SPDX-License-Identifier: Apache-2.0
+from typing import List, Optional, Tuple
+
+import torch
+import torch.fx as fx
+from torch._inductor.pattern_matcher import (Match, PatternMatcherPass,
+                                             fwd_only, register_replacement)
+
+from vllm.config import CompilationConfig
+from vllm.distributed import get_tp_group, tensor_model_parallel_all_reduce
+from vllm.distributed.parallel_state import (
+    get_tensor_model_parallel_world_size)
+from vllm.logger import init_logger
+
+from .vllm_inductor_pass import VllmInductorPass
+
+logger = init_logger(__name__)
+
+
+def search_embedding_all_reduce_rmsnorm(
+    arg2_1: torch.Tensor,
+    mul_6: torch.Tensor,
+    unsqueeze: torch.Tensor,
+    full_default: torch.Tensor,
+    permute: torch.Tensor,
+    arg3_1: torch.Tensor,
+):
+    embedding = torch.ops.aten.embedding.default(arg2_1, mul_6)
+    where = torch.ops.aten.where.self(unsqueeze, full_default, embedding)
+    all_reduce = tensor_model_parallel_all_reduce(where)
+    rmsnorm = torch.ops.higher_order.auto_functionalized(
+        torch.ops._C.rms_norm.default,
+        result=permute,
+        input=all_reduce,
+        weight=arg3_1,
+        epsilon=1e-5,
+    )
```

Comment:
It seems that this only matches if we use the custom rmsnorm op.

From the summary in this PR, and from trying it myself, I see that `custom_ops=["+rms_norm"],` is necessary for this pass to do any fusion. Makes me think the passes are too brittle and we may want to return to that "noop sentinel" idea we were discussing, @yaochengji

Reply:
In my understanding, with `custom_ops=["+rms_norm"]`, it is wrapped as a higher_order op. And we only need to match the same higher_order op for different hardware implementations.

Reply:
This could be parametrized so that any rms_norm implementation (torch-based or custom) works here. Unless RMS doesn't need to be a part of the pattern?

Reply:
I tried using the torch based op here and it broke the pattern in other ways.  IIRC it was because pytorch did some CSE-like rewrites involving on one of the values used by the torch based rmsnorm.

Reply:
I see, then I guess sentinel it is?